### PR TITLE
Feature/146 ingame upgrade system

### PR DESCRIPTION
**Pull request by KevinGit9**
Closes #<issue-number>
### Summary
Adds the upgrade system for in a run

### 1. Merging
- [x] Did you select the dev branch?
### 2. File Structure
- [x] Are all scripts placed within Assets/Scripts and organized into logical categories using PascalCase (e.g., Core, Player, Level)?
- [x] Are assets like materials, textures and sound files named using lower_snake_case (e.g., diamond_pickaxe.png)?
- [x] Did the changes avoid unnecessary scene modifications? If new changes were needed for testing, were they done in Scenes/Dev/<Specification> and then discarded?
### 3. Code
- [x] Do private fields start with an underscore and use camelCase (e.g., _privateVariable)?
- [x] Are static readonly variables named using UPPER_SNAKE_CASE?
- [x] Are enum types also in UPPER_SNAKE_CASE?
- [x] Are function and variable names clear, concise, and descriptive without using unnecessary abbreviations?
- [x] Are all public functions and fields documented with clear summaries where needed
- [x] Does the code maintain a strict hierarchy, ensuring objects do not fetch references themselves or their parents in the hierarchy?
- [x] Are references passed down through dependency injection as needed?Does each component only depend on its parent in the hierarchy?
- [x] Have unnecessary/testing Debug.Logs been removed?
- [x] Has commented code been removed?
### 4. Logging
- [] If possible, did you make a video and did you send it/a link to it in the feature-videos channel?

### CODE REVIEW

File: Project/Assets/Scripts/Upgrades/UpgradeData.cs [hunk 1]
```diff
@@ -1,19 +1,39 @@
...
 public class UpgradeData : ScriptableObject
 {
-    public string UpgradeName;
+    public string Name;
     public Sprite Icon;
     public string Description;
     public UpgradeType Type;
-    public float Value;
+    public UpgradeRarity Rarity;
+    public int MaxAmount = 5;
```
Comment (IcyKoala) [inline Project/Assets/Scripts/Upgrades/UpgradeData.cs:21]: Is there a reason this has a default value?

File: Project/Assets/Scripts/Upgrades/UpgradeDisplay.cs [hunk 2]
```diff
@@ -6,19 +6,25 @@
...
+    public RarityBracket CurrentRarityBracket;
+
     public List<UpgradeCard> CardSlots;
     public List<UpgradeData> AllUpgrades;
-    private List<UpgradeData> _availableUpgrades = new List<UpgradeData>();
-    private Dictionary<UpgradeData, int> _upgradeCounts = new Dictionary<UpgradeData, int>();
     public bool PoolNeedsUpdate = true;
+    private Dictionary<UpgradeRarity, List<UpgradeData>> _upgradePool = new();
+    private Dictionary<UpgradeData, int> _upgradeCounts = new Dictionary<UpgradeData, int>();
+    private UIManager _uIManager;
```
Comment (IcyKoala) [inline Project/Assets/Scripts/Upgrades/UpgradeDisplay.cs:18]: This follows the naming convention, but it just feels cursed ;-; (had to mention it)

File: Project/Assets/Scripts/Upgrades/UpgradeData.cs (hunk 1 above)
Comment (KevinGit9) [reply Project/Assets/Scripts/Upgrades/UpgradeData.cs:21]: Yes, cuz i agreed with Wout that the default cap for upgrades should be 5 for now, and those that are unique for example would just get a MaxAmount of 1.

File: Project/Assets/Scripts/Upgrades/Effects/AttackModifier.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+using UnityEngine;
+
+[CreateAssetMenu(fileName = "AttackModifier", menuName = "Upgrades/Effects/AttackModifier")]
+public class AttackModifier : UpgradeEffect
+{
+    private PlayerComponentManager _playerComponentManager;
+
+    public override void ApplyEffect(float effectValue)
+    {
+        _playerComponentManager = GameManager.Player.GetComponent<PlayerComponentManager>();
```
[the remark this answers is not in the document]
Comment (KevinGit9) [reply Project/Assets/Scripts/Upgrades/Effects/AttackModifier.cs]: Constructor doesnt work. It isnt called by unity when initializing a scriptable object. I've changed it to be lazy loaded instead so it doesnt fetch it every time.

File: Project/Assets/Scripts/Managers/RoomManager.cs [hunk 4]
```diff
@@ -37,51 +41,72 @@ void Start()
...
-    public void IncreaseRoomClear(){
+    public void IncreaseRoomClear()
+    {
         _roomClearCount++;
-        if(_roomClearCount % DifficultyScale == 0){
+        if (_roomClearCount % DifficultyScale == 0)
+        {
             DifficultyManager.IncreaseDifficulty();
         }
+        if (_roomClearCount % 5 == 0)
```
Comment (IcyKoala) [inline Project/Assets/Scripts/Managers/RoomManager.cs:54]: Prolly make this a stat similar to DifficultyScale

File: Project/Assets/Scripts/Managers/UIManager.cs
Comment (IcyKoala) [inline Project/Assets/Scripts/Managers/UIManager.cs:1]: i approve of these changes

**Review by IcyKoala — APPROVED** [review]
ty for the changes

**Pull request merged**